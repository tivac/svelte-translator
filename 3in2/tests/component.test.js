"use strict";

const wait = require("p-immediate");

describe("3in2 component wrapper", () => {
    const Wrapper = require("../component.html");
    const { default : ComponentA } = require("./specimens/component-a.svelte");
    const { default : ComponentB } = require("./specimens/component-b.svelte");

    let root;

    beforeEach(() => {
        root = document.createElement("div");
    });

    it("should render a svelte3 component", async () => {
        new Wrapper({
            target : root,

            data : {
                component : ComponentA,
            },
        });

        expect(root.innerHTML).toMatchSnapshot();
    });

    it("should throw if not given a svelte3 component", async () => {
        expect(() => new Wrapper({
            target : root,
            data   : {
                component : true
            },
        })).toThrowErrorMatchingSnapshot();
    });

    it("should destroy the svelte3 component when the wrapper is destroyed", async () => {
        const wrapper = new Wrapper({
            target : root,

            data : {
                component : ComponentA,
            },
        });

        wrapper.destroy();

        expect(wrapper.instance).toBeNull();
    });

    it("should render a svelte3 component w/ props", async () => {
        new Wrapper({
            target : root,

            data : {
                component : ComponentA,

                a : "A",
            },
        });

        expect(root.innerHTML).toMatchSnapshot();
    });

    it("should update the svelte3 component when props change", async () => {
        const wrapper = new Wrapper({
            target : root,

            data : {
                component : ComponentA,

                a : "A",
            },
        });

        expect(root.innerHTML).toMatchSnapshot();

        wrapper.set({
            a : "A2",
        });

        await wait();

        expect(root.innerHTML).toMatchSnapshot();
    });

    it("shouldn't recreate the svelte3 component when props change", async () => {
        const wrapper = new Wrapper({
            target : root,

            data : {
                component : ComponentA,

                a : "A",
            },
        });

        const child = wrapper.instance;

        wrapper.set({
            component : ComponentA,
            a : "A2",
        });

        await wait();

        expect(child).toBe(wrapper.instance);
    });

    it("should support swapping out the svelte3 component", async () => {
        const wrapper = new Wrapper({
            target : root,

            data : {
                component : ComponentA,

                a : "A",
            },
        });

        expect(root.innerHTML).toMatchSnapshot();

        wrapper.set({
            component : ComponentB,
        });

        await wait();

        expect(root.innerHTML).toMatchSnapshot();
    });

    it("should support creation without a component", async () => {
        const wrapper = new Wrapper({
            target : root,
        });

        expect(root.innerHTML).toMatchSnapshot();

        wrapper.set({
            component : ComponentB,
        });

        await wait();

        expect(root.innerHTML).toMatchSnapshot();
    });

    it("should set attributes on the wrapper element it creates", async () => {
        new Wrapper({
            target : root,

            data : {
                component : ComponentA,

                attrs : {
                    class      : "class",
                    style      : "color: red;",
                    "data-foo" : "data-foo",
                }
            },
        });

        expect(root.innerHTML).toMatchSnapshot();
    });
});
