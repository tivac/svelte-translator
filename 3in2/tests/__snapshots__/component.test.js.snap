// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`3in2 component wrapper should render a svelte3 component 1`] = `

<div>
  <p>
    The answer is a mystery
  </p>
</div>

`;

exports[`3in2 component wrapper should render a svelte3 component w/ props 1`] = `

<div>
  <p>
    The answer is 42
  </p>
</div>

`;

exports[`3in2 component wrapper should set attributes on the wrapper element it creates 1`] = `

<div class="class"
     style="color: red;"
     data-foo="data-foo"
>
  <p>
    The answer is a mystery
  </p>
</div>

`;

exports[`3in2 component wrapper should update the svelte3 component when props change 1`] = `

<div>
  <p>
    The answer is 42
  </p>
</div>

`;

exports[`3in2 component wrapper should update the svelte3 component when props change 2`] = `

<div>
  <p>
    The answer is science
  </p>
</div>

`;
